torch_geometric/nn/conv/disen_conv.py

```diff
@@ -0,0 +1,155 @@
+
+        assert K > 0
+        assert T > 0
+        assert tao > 0
+
+        self.T = T
+        self.tao = tao
+        self.in_channels = in_channels
+        self.out_channels = out_channels
+
+        if separate_channels is None:
+            if K is None:
+                raise ValueError(
+                    "`K` must be assigned when `separate_channels` is None")
+            self.separate_channels = [out_channels // K] * K
+            self.separate_channels[-1] += out_channels % K
+            self.K = K
+        else:
+            assert torch.all(torch.tensor(separate_channels) > 0)
+            assert sum(separate_channels) == out_channels
+            self.separate_channels = separate_channels
+            self.K = len(separate_channels)
+
+        self.lin_projector = Linear(
+            in_channels,
+            out_channels,
+            bias=True,
+            weight_initializer=weight_initializer,
+            bias_initializer=bias_initializer,
+        )
+
+    def reset_parameters(self):
+        self.lin_projector.reset_parameters()
+
+    def get_projector(
+        self, separate=False
+    ) -> Tuple[Union[Tuple[Tensor], Tensor], Union[Tuple[Tensor], Tensor]]:
+        projector = self.lin_projector.weight
+        projector_bias = self.lin_projector.bias
+        if separate:
+            projector = torch.split(projector, self.separate_channels, dim=0)
+            projector_bias = torch.split(projector_bias,
+                                         self.separate_channels, dim=0)
+        return projector, projector_bias
+
+    def project2aspects(self, features: Tensor) -> List[Tensor]:
+        Z = self.lin_projector(features)
+        Z = torch.split(Z, self.separate_channels, dim=-1)  # tuple of tensors
+        zs = [ck / ck.norm(dim=1, keepdim=True) for ck in Z]
+        return zs
+
+    def update_p(
+        self,
+        zs: List[Tensor],
+        cs: List[Tensor],
+        dst_nodes: Union[Tensor, List[int]],
+        src_nodes: Union[Tensor, List[int]],
+    ) -> Tensor:
+        assert len(dst_nodes) == len(src_nodes)
+        pout = [
+            torch.sum(cs[k][dst_nodes] * zs[k][src_nodes],
+                      dim=-1).div_(self.tao) for k in range(self.K)
+        ]
+        pout = F.softmax(torch.stack(pout), dim=0)
+        return pout
+
+    def forward(self, x: Tensor, edge_index: Union[SparseTensor, Tensor],
+                batch_size: int = None) -> Tensor:
+
+        if isinstance(edge_index, Tensor):
+            src_nodes, dst_nodes = edge_index
+        elif isinstance(edge_index, SparseTensor):
+            dst_nodes, src_nodes, _ = edge_index.coo()
+        else:
+            raise TypeError(f"{type(edge_index)} not a valid graph type")
+        bs = batch_size if batch_size is not None \
+            else dst_nodes.unique().numel()
+
+        Z = self.project2aspects(x)
+        # target nodes always placed first
+        C = [chunk[:bs].clone() for chunk in Z]
```

Comment:
Not really sure why we need to slice here based on `batch_size`. Can you explain?

Reply:
Assume there are `K` Disentangled factors responsible for `d1,d2,...,dK` output channels separately.  We have `batch_size` target nodes and the induced `1-hop` subgraph with totally `N` nodes.

`Z=[Z1(N, d1), Z2(N, d2), ...,  ZK(N, dK)]` is a list of `K` aspects of `N` nodes.  During the routing iteration described in the following figure, we need to reserve the entire `Z` for message passing and initialize the `K` segments of output of target nodes with `Zk[:batch_szie]`s.  `C` is alongside cloned to simultaneously allocate memory for the final disentangled output and initialize them. 

 
<img width="507" alt="屏幕截图 2022-04-06 154146" src="https://user-images.githubusercontent.com/26434203/161922158-733288f7-d85f-4160-a018-39642418efb1.png">

Reply:
## 1. Input Tuple (without `batch_size`)
If we input a tuple `(x_src, s_dst)`, the code would be like:
```python

def forward_bipartite(self, x: OptPairTensor, edge_index: Union[SparseTensor, Tensor], size: Size = None) -> Tensor:
        if isinstance(edge_index, Tensor):
            src_nodes, dst_nodes = edge_index
        elif isinstance(edge_index, SparseTensor):
            dst_nodes, src_nodes, _ = edge_index.coo()
        else:
            raise TypeError(f"{type(edge_index)} not a valid graph type")

        x_src, x_dst = x
        bipartite_size = (x_src.shape[0] + x_dst.shape[0], x_dst.shape[0]) if size is None else size

        z_dst = self.project2aspects(x_dst)
        z_src = self.project2aspects(x_src)
        z_all = [torch.cat([z_dst[k], z_src[k]]) for k in range(self.K)]  ## 1st marker 

        c_dst = [ck.clone() for ck in z_dst]

        for i in range(self.T):
            p = self.update_p(z_all, c_dst, dst_nodes, src_nodes)
            for k in range(self.K):
                graph = (
                    edge_index.set_value(p[k], layout="coo")
                    if isinstance(edge_index, SparseTensor)
                    else edge_index
                )
                ck = ( ## 2nd marker 
                        self.propagate(graph, x=(z_all[k], z_dst[k]), edge_weight=p[k],  size=bipartite_size)+ z_dst[k]
                )
                c_dst[k] = ck / ck.norm(dim=-1, keepdim=True)

        out = torch.cat([c_dst[k] for k in range(self.K)], dim=1)
        return out
```

This leads to a concatenation operation in the `1st marker` line. To avoid the allocation of new memory caused by `torch.cat`, it would be better to input `(x_all, x_dst)` to bypass `cat`. 

## 2. Input All (with `batch_size`) like the last commit 
If we input `x` of all nodes and the target nodes are first placed, `batch_szie` is required to locate the target nodes.

Actually, I prefer the solution of inputting `(x_all, x_dst)` without `batch_size` now.

Reply:
Thanks for clarification. I don't think we need to do anything with `batch_size` at all. The input matrix `x` is used for both central node features and features of neighbors such that `c_k` is just given via `Z[k]`, right? We can then calcuate `p_{v,k}` directly inside message (by querying the central node `c_i` and its neighbors `z_j`).

Reply:
Thanks for the comments. 

> The input matrix `x` is used for both central node features and features of neighbors such that `c_k` is just given via `Z[k]`?


Not exactly correct. `c_k` is initialized with `Z[k][dst_nodes]` but going to deviate from this fixed initialization value during the following iteration.  So we have to preserve both `Z[k]` and `c_k`.

I think it might be not that good to calculate `p_{v,k}` during the propagation. The message propagation is conducted for all `K` factors. For the propagation of `k`-th factor, only the inner product of `z_{v,k}` and `c_{k}` is computed, but  `p_{v,k}` is obtained via a softmax operation across all `K` such inner products.  This is why I prefer to compute all   `p_{v,k}` s before propagation.

Reply:
Sounds good. In this case, we should go for the input tuple approach as illustrated by your earlier. Importantly, we can also support `x` as a `torch.Tensor` via:
```python
if isinstance(x, Tensor):
    x = (x, x)
    
x_src, x_dst = x
```

Reply:
OK. I will commit later. 😃